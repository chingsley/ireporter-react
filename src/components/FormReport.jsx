/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createNewReport, editReport } from '../actions/report';
import setPreviewImage from '../utilities/setPreviewImage';

export class Report extends Component {
  constructor(props) {
    super(props);
    const { report } = props;
    const {
      id, location, comment, status, type, Images,
    } = report;

    this.state = {
      // address: '',
      id: report ? id : null,
      location: report ? location : '',
      comment: report ? comment : '',
      status: report ? status : 'draft',
      type: report ? type : 'red-flag',
      images: (report && Images && Images[0]) ? Images[0] : null,
      previewImg: (report && Images && Images[0]) ? Images[0] : null,
    };

    this.history = props.history;
    this.onSubmit = this.onSubmit.bind(this);
    this.onImgChange = this.onImgChange.bind(this);
  }

  async onSubmit(e) {
    e.preventDefault();

    const { mode } = this.props;
    if (mode === 'create') {
      const { isLoggedIn, history } = this.props;
      if (!isLoggedIn) {
        history.push('/login');
      } else {
        const { createNewReport: reportCase } = this.props;
        await reportCase({ ...this.state });
      }
    }

    if (mode === 'edit') {
      const { editReport: edit } = this.props;
      await edit({ ...this.state });
    }
  }

  onImgChange(e) {
    return this.setState({ images: e.target.files[0], previewImg: e.target.files });
  }


  render() {
    const {
      // address,
      location,
      comment,
      previewImg,
    } = this.state;
    const { mode } = this.props;
    return (
      <form
        onSubmit={this.onSubmit}
        className="report-form"
        id="report-form"
      >
        {mode === 'create'
        && (
          <div className="report-form__radio-container">
            <div className="report-form__radio-container--inner-container">
              <input
                type="radio"
                className="report-form__radio-container-radio"
                id="redflag"
                name="type"
                value="red-flags"
                onClick={e => this.setState({ type: e.target.value })}
                required
              />redflag
            </div>
            <div className="report-form__radio-container--inner-container">
              <input
                type="radio"
                className="report-form__radio-container-radio"
                id="intervention"
                name="type"
                value="interventions"
                onClick={e => this.setState({ type: e.target.value })}
                required
              />intervention
            </div>
          </div>
        )}
        <input
          type="text"
          className="report-form__input"
          id="coords"
          placeholder="coordinates"
          value={location}
          onChange={e => this.setState({ location: e.target.value })}
          required
        />
        <textarea
          name="comment"
          className="report-form__input-textarea"
          id="comment"
          rows="6"
          placeholder="comments"
          value={comment}
          onChange={e => this.setState({ comment: e.target.value })}
          required
        />
        {mode === 'create'
          && (
          <>
            <input
              className="report-form-image-upload-input"
              type="file"
              name="image"
              onChange={this.onImgChange}
              ref={fileInput => this.fileInput = fileInput}
            />
            <div className="form-report-image-container" onClick={() => this.fileInput.click()}>
              <span className="form-report-image-container__add-image">Add Image</span>
              <img
                className="form-report-image-container__selected-image"
                src={setPreviewImage(previewImg)}
                alt="report-evidence"
              />
            </div>
            <button
              className="report-form__button"
              id="btn-send-report"
              type="submit"
            >
              SEND
            </button>
          </>
          )}
        {mode === 'edit'
          && (
            <button
              className="report-form__button"
              id="btn-send-report"
              type="submit"
            >
                SAVE CHANGES
            </button>
          )}
      </form>
    );
  }
}

Report.defaultProps = {
  isLoggedIn: false,
  mode: 'create',
  report: {},
};

Report.propTypes = {
  createNewReport: PropTypes.func.isRequired,
  editReport: PropTypes.func.isRequired,
  isLoggedIn: PropTypes.bool,
  history: PropTypes.oneOfType([PropTypes.object, PropTypes.func]).isRequired,
  mode: PropTypes.string,
  report: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
};

const mapStateToProps = state => ({
  isLoggedIn: state.auth.isLoggedIn,
  fetching: state.fetchStatus.fetching,
});

const FormReport = connect(mapStateToProps, { createNewReport, editReport })(Report);
export default FormReport;
