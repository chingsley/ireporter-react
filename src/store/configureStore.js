import {
  applyMiddleware, createStore, combineReducers, compose,
} from 'redux';
import thunk from 'redux-thunk';
import authReducer from '../reducers/authReducer';

export default () => {
  const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // eslint-disable-line
  const store = createStore(
    combineReducers({
      auth: authReducer,
    }),
    composeEnhancers(applyMiddleware(thunk)),
  );
  return store;
};
