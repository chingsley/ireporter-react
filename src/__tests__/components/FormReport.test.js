import React from 'react';
import { shallow, mount } from 'enzyme';
// import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import FormReport from '../../components/FormReport';

const wrap = () => {
  const props = {
    history: {
      push: jest.fn(),
      location: {
        pathname: '',
      },
    },

    createReport: jest.fn(),
    isLoggedIn: false,
    fetching: false,
  };

  const mockStore = configureStore([thunk]);

  const store = mockStore({
    auth: {
      isLoggedIn: true,
      user: {},
    },
    fetchStatus: {
      fetching: false,
      errMessage: '',
    },
    report: {
      report: {},
    },
  });

  const shallowWrapper = shallow(<FormReport {...props} />);

  const mountWrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <FormReport {...props} />
      </MemoryRouter>
    </Provider>,
  );
  return { shallowWrapper, mountWrapper };
};


describe('Test the report form', () => {
  const { mountWrapper: wrapper } = wrap();
  it('should render with the correct elements', () => {
    expect(wrapper.find('form.report-form').exists()).toBe(true);
    expect(wrapper.find('div.report-form__radio-container').exists()).toBe(true);
    expect(wrapper.find('div.report-form__radio-container--inner-container').exists()).toBe(true);
    expect(wrapper.find('input.report-form__radio-container-radio').exists()).toBe(true);
    expect(wrapper.find('div.report-form__radio-container--inner-container').exists()).toBe(true);
    expect(wrapper.find('input.report-form__radio-container-radio').exists()).toBe(true);
    expect(wrapper.find('input.report-form__input').exists()).toBe(true);
    expect(wrapper.find('textarea.report-form__input-textarea').exists()).toBe(true);
    expect(wrapper.find('button.report-form__button').exists()).toBe(true);
  });
});
