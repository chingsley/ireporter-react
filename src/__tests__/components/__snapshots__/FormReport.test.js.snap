// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the report form should render with the correct elements 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "pnhtgl",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "pnhtgl",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Report)
        createReport={[MockFunction]}
        fetching={false}
        history={
          Object {
            "location": Object {
              "pathname": "",
            },
            "push": [MockFunction],
          }
        }
        isLoggedIn={false}
      >
        <Report
          createNewReport={[Function]}
          createReport={[MockFunction]}
          editReport={[Function]}
          fetching={false}
          history={
            Object {
              "location": Object {
                "pathname": "",
              },
              "push": [MockFunction],
            }
          }
          isLoggedIn={true}
          mode="create"
          report={Object {}}
        >
          <form
            className="report-form"
            id="report-form"
            onSubmit={[Function]}
          >
            <div
              className="report-form__radio-container"
            >
              <div
                className="report-form__radio-container--inner-container"
              >
                <input
                  className="report-form__radio-container-radio"
                  id="redflag"
                  name="type"
                  onClick={[Function]}
                  required={true}
                  type="radio"
                  value="red-flags"
                />
                <label
                  htmlFor="redflag"
                  id="label-redflag"
                >
                  redflag
                </label>
              </div>
              <div
                className="report-form__radio-container--inner-container"
              >
                <input
                  className="report-form__radio-container-radio"
                  id="intervention"
                  name="type"
                  onClick={[Function]}
                  required={true}
                  type="radio"
                  value="interventions"
                />
                <label
                  htmlFor="intervention"
                  id="label-intervention"
                >
                  intervention
                </label>
              </div>
            </div>
            <input
              className="report-form__input"
              id="coords"
              onChange={[Function]}
              placeholder="coordinates"
              required={true}
              type="text"
            />
            <textarea
              className="report-form__input-textarea"
              id="comment"
              name="comment"
              onChange={[Function]}
              placeholder="comments"
              required={true}
              rows="6"
            />
            <input
              className="report-form-image-upload-input"
              name="image"
              onChange={[Function]}
              type="file"
            />
            <div
              className="form-report-image-container"
              onClick={[Function]}
            >
              <span
                className="form-report-image-container__add-image"
              >
                Add Image
              </span>
              <img
                alt="report-evidence"
                className="form-report-image-container__selected-image"
                src="https://via.placeholder.com/700x400"
              />
            </div>
            <button
              className="report-form__button"
              id="btn-send-report"
              type="submit"
            >
              SEND
            </button>
          </form>
        </Report>
      </Connect(Report)>
    </Router>
  </MemoryRouter>
</Provider>
`;
